climada/engine/supplychain.py

```diff
@@ -0,0 +1,322 @@
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+---
+
+Define the SupplyChain class.
+"""
+
+__all__ = ['SupplyChain']
+
+import logging
+import datetime as dt
+from pathlib import Path
+from tqdm import tqdm
+import numpy as np
+import pandas as pd
+from iso3166 import countries_by_numeric
+from climada.util import files_handler as u_fh
+from climada.engine import Impact
+from climada.util.constants import SYSTEM_DIR
+from climada.entity.exposures.base import Exposures
+
+LOGGER = logging.getLogger(__name__)
+WIOD_FILE_LINK = 'http://www.wiod.org/protected3/data16/wiot_ROW/'
+
+class SupplyChain():
+    """SupplyChain class.
+
+    The SupplyChain class provides methods for loading input-output tables and
+    computing direct, indirect and total impacts.
+
+    Attributes
+    ----------
+    mriot_data : np.array
+        The input-output table data.
+    countries_iso3 : np.array
+        Countries considered by the input-output table reported via ISO3 codes.
+    sectors : np.array
+        Sectors considered by the input-output table.
+    total_prod : np.array
+        Countries' total production.
+    mriot_type : str
+        Type of the adopted input-output table. Currently only WIOD tables
+        are supported.
+    cntry_pos : dict
+        Countrie's positions within the input-output tables and impact arrays.
+    cntry_dir_imp : list
+        Countries undergoing direct impacts.
+    years : np.array
+        Years of the considered hazard events for which impact is calculated.
+    direct_impact : np.array
+        Direct impact array.
+    direct_aai_agg : np.array
+        Average annual direct impact array.
+    indirect_impact : np.array
+        Indirect impact array.
+    indirect_aai_agg : np.array
+        Average annual indirect impact array.
+    total_impact : np.array
+        Total impact array.
+    total_aai_agg : np.array
+        Average annual total impact array.
+    io_data : dict
+        Dictionary with the coefficients, inverse and risk_structure matrixes and
+        the selected input-output approach.
+    """
+
+    def __init__(self):
+        """Initialize SupplyChain."""
+        self.mriot_data = np.array([], dtype='f')
+        self.countries_iso3 = np.array([], dtype='str')
+        self.sectors = np.array([], dtype='str')
+        self.total_prod = np.array([], dtype='f')
+        self.mriot_type = ''
+        self.cntry_pos = {}
+        self.years = np.array([], dtype='f')
+        self.direct_impact = np.array([], dtype='f')
+        self.direct_aai_agg = np.array([], dtype='f')
+        self.indirect_impact = np.array([], dtype='f')
+        self.indirect_aai_agg = np.array([], dtype='f')
+        self.total_impact = np.array([], dtype='f')
+        self.total_aai_agg = np.array([], dtype='f')
+        self.io_data = {}
+
+    def read_wiod16(self, year=2014, file_folder=SYSTEM_DIR.joinpath('results')):
```

Comment:
Although it's probably fine, function calls in optional parameter default assignment make me uneasy. I'd prefer `..., file_folder=None):` and then in the method block:
``` if file_folder = file_folder or SYSTEM_DIR.joinpath('results')```

Is `climada/system/results` really the preferred path for WIOD tables, which are basically input data?

Actually - wouldn't it be appropriate to have a folder `climada/system/wiod` just fort this? In this case you can also remove the parameter altogether: wiod data goes there. period.
You could use the config file to specify it:
"engine": {
        "supplychain": {
            "resources": {
                "wiod_16": "http://www.wiod.org/protected3/data16/wiot_ROW/"
            },
            "local_data": "{local_data.system}/WIOD"
        }
    },

Only need to make sure the test file is found during test. Either copy it over there or, even better, overwrite the config values.

Reply:
Thanks @emanuel-schmid. This is exactly the kind of tip I was looking for. I am only missing how to do the last, i.e., make sure that the test files are found under `CONFIG.engine.supplychain.resources.wiod_16.str()`

ps
the test files **cannot** be downloaded

Reply:
Suggestions were well received and implemented, thanks. I am just unsure whether I am doing the below effectively:

> Only need to make sure the test file is found during test. Either copy it over there or, even better, overwrite the config values.

as I copy the file from the test folder to the system/WIOD folder using shutil.copy (see line 42 of test_supplychain.py in commit https://github.com/CLIMADA-project/climada_python/pull/180/commits/ed468290bf322d343c230549529e51c09832e9d0)

other/better ways are welcome; please explicit code in case.